examples/pycaffe/layers/pascal_multilabel_datalayers.py

```diff
@@ -0,0 +1,258 @@
+
+    def dispatch_worker(self):
+        assert self.thread is None
+        self.thread = Thread(target=self.batch_advancer)
+        self.thread.start()
+
+    def join_worker(self):
+        assert self.thread is not None
+        self.thread.join()
+        self.thread = None
+
+    def backward(self, top, propagate_down, bottom):
+        """ this layer does not back propagate """
+        pass
+
+
+class BatchAdvancer():
+    """
+    This is the class that is run asynchronously and actually does the work.
+    """
+    def __init__(self, result, params):
+        self.result = result
+        self.batch_size = params['batch_size'] 
+        self.im_shape = params['im_shape'] 
+        self.pascal_root = params['pascal_root']
+        self.im_shape = params['im_shape']
+        self.indexlist = [line.rstrip('\n') for line in open(osp.join(self.pascal_root, 'ImageSets/Main', params['split'] + '.txt'))] #get list of image indexes.
+        self._cur = 0 # current image
+        self.transformer = SimpleTransformer() #this class does some simple data-manipulations
+
+        print "BatchAdvancer initialized with {} images".format(len(self.indexlist))
+
+    def __call__(self):
+        """
+        This does the same stuff as the forward layer of the synchronous layer. Exept that we store the data and labels in the result dictionary (as lists of length batchsize).
+        """
+        self.result['data'] = []
+        self.result['label'] = []
+        for itt in range(self.batch_size):
+
+            # Did we finish an epoch?
+            if self._cur == len(self.indexlist):
+                self._cur = 0
+                shuffle(self.indexlist)
+            
+            # Load an image
+            index = self.indexlist[self._cur] # Get the image index
+            im = np.asarray(Image.open(osp.join(self.pascal_root, 'JPEGImages', index + '.jpg'))) # load image
+            im = scipy.misc.imresize(im, self.im_shape) # resize
+            
+            # do a simple horizontal flip as data augmentation
+            flip = np.random.choice(2)*2-1
+            im = im[:, ::flip, :]
+
+            # Load and prepare ground truth
+            multilabel = np.zeros(20).astype(np.float32)
+            anns = load_pascal_annotation(index, self.pascal_root)
+            for label in anns['gt_classes']:
+                # in the multilabel problem we don't care how MANY instances there are of each class. Only if they are present.
+                multilabel[label - 1] = 1 # The "-1" is b/c we are not interested in the background class.
+
+            # Store in a result list.
+            self.result['data'].append(self.transformer.preprocess(im))
+            self.result['label'].append(multilabel)
+            self._cur += 1
+
+
+def load_pascal_annotation(index, pascal_root):
+    """
+    This code is borrowed from Ross Girshick's FAST-RCNN code. It parses the PASCAL .xml metadata files.
+    Thanks Ross!
```

Comment:
You should cite the URL (https://github.com/rbgirshick/fast-rcnn) and tech report (http://arxiv.org/abs/1504.08083) that accompanies Fast R-CNN.  The current citation isn't quite enough.